README.md

```diff
@@ -45,102 +53,241 @@ console.log(result);
+
+There is, however, no blanket rule for what to do; each error represents a specific problem in the context of an entire application and the appropriate response to it will be heavily context dependent.
+
+## Approaches
+Good error handling is typically not something that can just be bolted onto an existing program as an afterthought. Well conceived error handling will affect the structure of the code. In JavaScript and Node.js, there are a number of approaches, some of which are explored below.
+
+### General Principles
+Regardless of the chosen approach, there are some principles which can be generally applied:
+1. **Be consistent, not ad-hoc.**
+  * Inconsistent approaches to error handling will complicate your code and make it much harder to reason about.
+2. **Try to trip into a failure code path as early as possible.**
+  * For example, it may be tempting to return default values in the case of an error and allow the application to continue as normal.
+  * This may be appropriate in some cases, but can often cover up the root cause of an error and make it difficult to track down, or result in unhelpful error messages.
+3. **Propagate errors to a part of the application that has sufficient context to know how to deal with them.**
+  * Many times, we will write generic functions to perform common actions, like making a network request.
+  * If the network request fails, the generic function cannot infer the appropriate response, because it doesn't know which part of the application it has been called from.
+  * It should therefore try to propagate the error to its caller instead of trying to recover directly.
+
+### Illustrative Example
+To illustrate the three approaches we will cover, we will use the same simple example in each, so that comparisons are easier. Imagine you intend to write a function `applyToInteger`, with the following signature:
+```js
+applyToInteger(func: Function, integer: Number)
+```
+That is, the function accepts two arguments, `func`, which is a `Function`, and `integer` which is a whole `Number`. It applies the `func` to `integer` and returns the result. We will use this example to explore how to deal with unexpected inputs.
 
-If **arrayOfNumbersIThink** was retrieved from an API call, we can't always be certain the values will be what we expect. How can we deal with these unknown situations?
+### Approach 1: Throwing and Catching Errors
 
-### 1. Error Throwing
+#### Throwing
+During runtime, errors can be thrown in our application unexpectedly by computations acting on faulty computations produced earlier (like the first example above). We can also manually throw errors ourselves by using the [`throw`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/throw) keyword. This will immediately terminate the application, unless there is a [`catch`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/try...catch) block in the call stack.
```

Comment:
Okay, I think one thing missing here is around throwing errors in async functions.

I am not sure how we can explain this without too much detail.
Basic idea: errors unwind the call stack, looking for a catch. Async functions have their own call stack. Placing a catch around an async function will not catch errors, and they will crash the application.

Not sure how much detail you want to go into. But I think there should be a little more emphasis not throwing from async functions / trying to catch around async.